components/Observer.vue

```diff
@@ -0,0 +1,32 @@
+<template>
+  <div class="observer" />
+</template>
+
+<script>
+export default {
+  props: {
+    options: {
+      type: Object,
+      default: null
+    }
+  },
+  data: () => ({
+    observer: null
+  }),
+  mounted() {
+    const options = this.options || {}
+    // Use array destructing on the entry argument i.e. entries[0]
+    this.observer = new IntersectionObserver(([entry]) => {
+      if (entry && entry.isIntersecting) {
+        this.$emit('intersect')
```

Comment:
Might be better to name the component, events etc by the purpose not the implementation? Intersects is a bit opaque.

Reply:
We could check at this point if the element is still in the DOM.  I worry that there might be a timing window between the observer firing and our callback, unless you know otherwise.  If it's not in the DOM, would it be better not to emit.

I had a method for that somewhere, probably another project, hops up the parent nodes to see if it ends at body though there will be better ways.

Reply:
Yeah agreed.  This was the renamed version as I originally called it toInfinityAndBeyond.  Better than that but could be more semantic and less technical sounding.

So are you saying that potentially the intersection could happen, that triggers the data fetch and by the time that data comes back they could have scrolled back up the page?  In which case there would be no point adding it to the DOM?

Reply:
From edwh:

I like that name.

Not quite:
- scroll down
- trigger api call
- scroll up
- scroll down
- trigger api call again
- first api call now completes
- other stuff starts happening
- second api call now completes

Feels like there will be some cases where that would cause bugs.

Reply:
The jobs section wouldn't have a problem as it just increments a variable so it doesn't matter which one completes first.

The other type will load data that it has fetched from an API.  I can't remember what the code does off the top of my head but I'm guessing it is some type of array push.  As JavaScript is single threaded (ignoring web workers) then it shouldn't matter which one returns first unless you are fetching the data in some kind of controlled order and displaying it as is rather than controlling the order in the view.  In C# you would probably use a mutex or something similar to lock the array whilst you were updating the data.  In JavaScript it should be safe.

Definitely worth adding to the list of tests though.  We can mock up some dummy functions that return at different times.  

This type of work would really benefit from having a unit test framework.  Nick was investigating that wasn't he?  Did he conclude anything?  I've always used Ava because it was the only one I could get working with Nuxt (and I think Nuxt recommend it) but I haven't looked for a while so there might now be others that work.

Reply:
I can imagine more complex cases where it would matter.  

- Scroll and becomes visible, API call 1 starts
- Scroll and becomes visible, API call 2 starts
- API call 1 returns
- New "context" A set in client component for use on next API call (this is how we page through data)
- Scroll and becomes visible, API call 3 starts with new context A
- API call 3 completes with even newer context B
- New context B set in client component
- API call 2 completes with context A again
- Context A set in client component
- Next call will use context A rather than context B.

Will that matter?  I don't want to find out.  We just want to knock all of this kind of thing on the head by not triggering anything doing anything new if we are waiting for the callback to complete via finished() or loaded() became visible.

I agree that this is an excellent case for UT.  We didn't get as far with Nick as adopting something, so go with what you know.

Reply:
Crazy busy day today so I'll read this properly later.

I think it's worth having a look around to see which test frameworks play well with Nuxt now.  It's probably been over a year since I've looked at new ones.  Ava is alright. Seems to do the job but it always felt like I was fighting it when trying to add mocking libraries etc.